burr/core/application.py

```diff
@@ -1174,6 +1184,22 @@ def iterate(
         :return: Each iteration returns the result of running `step`. This generator also returns a tuple of
             [action, result, current state]
         """
+        # This is a gentle warning for existing users
+        if self._adapter_set.async_hooks:
+            logger.warning(
+                "There are asynchronous hooks present in the application that will be ignored. "
+                "Please use .aiterate or .arun methods to have them executed. "
+                f"The application has following asynchronous hooks: {self._adapter_set.async_hooks} "
+            )
+
+        # Seems fair to raise this if everything is async but the app execution
+        if self._adapter_set.async_hooks and isinstance(self._builder, AsyncApplicationBuilder):
```

Comment:
IMO we should error out when the application is built with `.abuild()`, has async hooks, and we run it with sync methods, e.g. `.run()` since this is very easy to miss that the async hooks just don't get called.

Vice versa, if we have an async application with `.arun()` and sync hooks, I wouldn't error it out since we cannot guarantee that we every adapter has async support / user might be ok for some things to be blocking. Maybe a warning here is more appropriate?

Reply:
Yes, I agree -- this is backwards compatible. The second case is odd, actually, as you've pointed out. To make it more concrete, not every synchronous hook is incompatible in an async mode. Take, for example, the hook that prints "doing X step" before every step -- this is just going to be a sync hook, not an async hook, but it should work in an async setting. No need to build an async version of something like that, when it never has an await.

I think we shold maintain the design where `.arun()` allows sync + async hooks, `.run()` only allows sync hooks (and should break otherwise, althugh that's an edge case...). Also, think this simplifies?